kyo-prelude/shared/src/main/scala/kyo2/kernel/Trace.scala

```diff
@@ -0,0 +1,103 @@
+package kyo2.kernel
+
+import internal.*
+import java.util.Arrays
+import java.util.concurrent.atomic.AtomicInteger
+import scala.annotation.tailrec
+import scala.util.control.NoStackTrace
+
+class Trace(
+    private[kernel] var frames: Array[Frame],
+    private[kernel] var index: Int
+)
+
+object Trace:
+
+    private[kernel] def init: Trace = Trace(new Array[Frame](maxTraceFrames), 0)
+
+    abstract class Owner extends TracePool.Local:
+        final private var frames = new Array[Frame](maxTraceFrames)
+        final private var index  = 0
+
+        final private[kernel] def pushFrame(frame: Frame): Unit =
+            val idx = this.index
+            frames(idx & (maxTraceFrames - 1)) = frame
+            this.index = idx + 1
+        end pushFrame
+
+        final private[kernel] def saveTrace(): Trace =
+            val newTrace   = borrow()
+            val newFrames  = newTrace.frames
+            val copyLength = math.min(index, maxTraceFrames)
+
+            if index <= maxTraceFrames then
+                System.arraycopy(frames, 0, newFrames, 0, copyLength)
+            else
+                val splitIndex      = index & (maxTraceFrames - 1)
+                val firstPartLength = maxTraceFrames - splitIndex
+                System.arraycopy(frames, splitIndex, newFrames, 0, firstPartLength)
+                System.arraycopy(frames, 0, newFrames, firstPartLength, splitIndex)
+            end if
+            newTrace.index = Math.max(maxTraceFrames, index)
+            newTrace
+        end saveTrace
+
+        final private[kyo2] def copyTrace(trace: Trace): Trace =
+            val newTrace = borrow()
+            System.arraycopy(trace.frames, 0, newTrace.frames, 0, Math.min(trace.index, maxTraceFrames - 1))
+            newTrace.index = trace.index
+            newTrace
+        end copyTrace
+
+        final private[kyo2] def releaseTrace(trace: Trace): Unit =
+            release(trace)
```

Comment:
This will be called from a point in the fiber implementation where it's guaranteed that the trace won't be used anymore because the fiber that holds it has completed.